File: app/src/test/java/net/tlalka/android/fiszki/models/dto/LessonDtoTest.java
package net.tlalka.android.fiszki.models.dto;

import com.google.gson.Gson;
import net.tlalka.android.fiszki.models.dto.json.LessonDto;
import net.tlalka.android.fiszki.models.dto.json.WordDto;
import net.tlalka.android.fiszki.models.types.LevelType;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.util.Arrays;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;

public class LessonDtoTest {

    private Gson gson;

    @Before
    public void setup() {
        this.gson = new Gson();
    }

    @Test
    public void testParseJsonToLessonDto() throws IOException {

        // given
        InputStream stream = getClass().getResourceAsStream("/samples/lesson.json");
        Reader reader = new InputStreamReader(stream, "UTF-8");

        // when
        LessonDto dto = gson.fromJson(reader, LessonDto.class);

        // then
        assertNotNull(dto);
        assertEquals(LevelType.BEGINNER, dto.getLevel());
        assertEquals(3, dto.getName().getLanguages().size());
        assertEquals(2, dto.getWords().size());
    }

    @Test
    public void testParseLessonDtoToJson() throws IOException {

        // given
        LessonDto dto = new LessonDto();
        dto.setName(new WordDto());
        dto.setLevel(LevelType.ADVANCED);
        dto.setWords(Arrays.asList(new WordDto(), new WordDto()));

        // when
        String result = gson.toJson(dto);

        // then
        assertNotNull(result);
        assertFalse(result.isEmpty());
    }
}

File: app/src/test/java/net/tlalka/android/fiszki/model/dto/LessonDtoTest.kt
package net.tlalka.android.fiszki.model.dto

import com.google.gson.Gson
import net.tlalka.android.fiszki.model.dto.json.LessonDto
import net.tlalka.android.fiszki.model.dto.json.WordDto
import net.tlalka.android.fiszki.model.types.LevelType
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Test
import java.io.InputStreamReader

class LessonDtoTest {

    @Test
    fun testParseJsonToLessonDto() {

        // given
        val stream = javaClass.getResourceAsStream("/samples/lesson.json")
        val reader = InputStreamReader(stream, "UTF-8")

        // when
        val dto = Gson().fromJson(reader, LessonDto::class.java)

        // then
        assertNotNull(dto)
        assertEquals(LevelType.BEGINNER, dto.level)
        assertEquals(3, dto.name.languages.size.toLong())
        assertEquals(2, dto.words.size.toLong())
    }

    @Test
    fun testParseLessonDtoToJson() {

        // given
        val dto = LessonDto()
        dto.name = WordDto()
        dto.level = LevelType.ADVANCED
        dto.words = listOf(WordDto(), WordDto())

        // when
        val result = Gson().toJson(dto)

        // then
        assertNotNull(result)
        assertFalse(result.isEmpty())
    }
}
